src/misc/cross.cpp

```diff
@@ -715,6 +715,88 @@ std::string cfstr_to_string(CFStringRef source)
 	return lang;
 }
 
+#if defined (__linux__)
+
+#include <fcntl.h>
+
+int get_num_physical_cpus()
+{
+	constexpr int MaxCpus = 128;
+	bool core_ids[MaxCpus] = {};
+	int num_cores = 0;
+	for (int i = 0; i < MaxCpus; ++i) {
+		std::string path = "/sys/devices/system/cpu/cpu" + std::to_string(i) + "/topology/core_id";
+		int fd = open(path.c_str(), O_RDONLY);
+		if (fd != -1) {
+			char core_id_str[6] = {};
+			if (read(fd, core_id_str, 5) > 0) {
+				int core_id = atoi(core_id_str);
+				if (core_id >= 0 && core_id < MaxCpus && !core_ids[core_id]) {
+					core_ids[core_id] = true;
+					++num_cores;
+				}
+			}
+			close(fd);
+		}
+	}
+	return std::max(num_cores, 1);
+}
+
+#elif defined (MACOSX)
+
+#include <sys/sysctl.h>
+
+int get_num_physical_cpus()
+{
+	int num_cpus = 0;
+	size_t size = sizeof(num_cpus);
+	sysctlbyname("hw.physicalcpu", &num_cpus, &size, nullptr, 0);
+	return std::max(num_cpus, 1);
+}
+
+#elif defined (WIN32)
+
+#include <Windows.h>
+
+int get_num_physical_cpus()
+{
+	// Size of the buffer as input, gets modified on function call to size of return data
+	DWORD buffer_size = 10 * 1024 * 1024;
```

Comment:
Nah I overdid it :laughing:

On my system it was less than 100 bytes per CPU core.  Can probably drop it down to 100KB and it'll be fine even on a Threadripper.

Reply:
You could use your `MaxCpus` constant here too.

Reply:
Shrunk to 100KB.  `MaxCpus` isn't really helpful without knowing an exact size of the struct per CPU (and the size might even vary from system to system).  This will probably be enough for 1000 cores or so though.